File: langs/javascript/src/merger.ts
/**
 * Deep merge utility for DOTLYTE configuration layers.
 */

/**
 * Deep merge two objects. Values in `override` take precedence.
 */
export function deepMerge(
  base: Record<string, unknown>,
  override: Record<string, unknown>,
): Record<string, unknown> {
  const result = { ...base };

  for (const [key, value] of Object.entries(override)) {
    if (
      key in result &&
      result[key] !== null &&
      typeof result[key] === "object" &&
      !Array.isArray(result[key]) &&
      value !== null &&
      typeof value === "object" &&
      !Array.isArray(value)
    ) {
      result[key] = deepMerge(
        result[key] as Record<string, unknown>,
        value as Record<string, unknown>,
      );
    } else {
      result[key] = value;
    }
  }

  return result;
}

File: langs/rust/src/merger.rs
//! Deep merge utility.

use serde_json::Value;

/// Deep merge two JSON object maps. Values in `override_map` take precedence.
pub fn deep_merge(
    base: serde_json::Map<String, Value>,
    override_map: serde_json::Map<String, Value>,
) -> serde_json::Map<String, Value> {
    let mut result = base;

    for (key, value) in override_map {
        if let Some(existing) = result.get(&key) {
            if let (Value::Object(base_obj), Value::Object(over_obj)) = (existing, &value) {
                let merged = deep_merge(base_obj.clone(), over_obj.clone());
                result.insert(key, Value::Object(merged));
                continue;
            }
        }
        result.insert(key, value);
    }

    result
}
